libraries/AP_Scripting/applets/repl.lua

```diff
@@ -0,0 +1,365 @@
+-- Interactive REPL (read-evaluate-print-loop) for the Lua scripting engine
+-- accessible over serial, with line editing, history, and output formatting.
+
+-- 0-based index of Scripting protocol port to use, or nil to use MAVLink
+local PORT_IDX = 0
+local MAX_HISTORY = 50 -- number of lines of history to keep (must be >= 1)
+local VERSION = "v1.0" -- version is convenience for the user
+
+local port
+if PORT_IDX == nil then
+  port = require("mavport")
+else
+  port = serial:find_serial(PORT_IDX)
+end
+assert(port, "REPL scripting port not configured")
+
+-- scan through parameters to find our port and grab its baud rate
+do
+  local serial_info = ""
+  local baud = 115200
+  if PORT_IDX ~= nil then
+    local port_num = 0
+    while PORT_IDX >= 0 and port_num <= 9 do
+      local protocol = param:get(("SERIAL%d_PROTOCOL"):format(port_num))
+      port_num = port_num + 1
+      if protocol == 28 then PORT_IDX = PORT_IDX - 1 end
+    end
+    if PORT_IDX == -1 then -- correct port index found
+      port_num = port_num - 1
+      baud = param:get(("SERIAL%d_BAUD"):format(port_num)) or 115200
+      serial_info = (" on SERIAL%d, BAUD=%d"):format(port_num, baud)
+    end
+  end
+  -- if we can't find the right port, the baud probably does not matter
+  -- (e.g. CAN or network port)
+  port:begin(baud)
+
+  gcs:send_text(6, "Lua REPL "..VERSION.." starting"..serial_info)
+end
+
+-- grab things we want from the environment in case the user messes them up
+local string = string
+local table = table
+local print_real = print
+
+-- declaration of main state variable and functions
+local state_func, state_read, state_eval, state_print
+
+-- write the string s to the port, buffering if not all could be written.
+-- buffer starts of with the first message and prompt
+local tx_buf = {"\r\n\r\nLua REPL "..VERSION.." started.\r\n> "}
+local writestring
+if port.writestring then -- use more efficient method if we have it
+  writestring = function(s)
+    if tx_buf then -- stuff already in the buffer?
+      tx_buf[#tx_buf+1] = s -- this needs to go after
+    else
+      local written = port:writestring(s)
+      if written < #s then
+        -- write was short i.e. system buffer is full. buffer the rest of the
+        -- string and transmit it later
+        tx_buf = { s:sub(written+1) }
+      end
+    end
+  end
+else
+  writestring = function(s)
+    if tx_buf then -- stuff already in the buffer?
+      tx_buf[#tx_buf+1] = s -- this needs to go after
+    else
+      for ci = 1, #s do
+        if port:write(s:byte(ci)) == 0 then
+          -- write failed i.e. system buffer is full
+          -- we now buffer the rest of the string and transmit it later
+          tx_buf = { s:sub(ci) }
+          break
+        end
+      end
+    end
+  end
+end
+
+-- substitute print function for within the REPL that prints to the port
+function print(...) -- luacheck: ignore
```

Comment:
What is this warning?

Reply:
Lua check doesn’t like shadowing a built in. I will add the number but I don’t think there’s a good way around the ignore.

Reply:
Fair enough, if you add the number and a short comment explaining why its needed I think its OK. I just want to minimize the number of things we have to ignore so we can prevent people copy and pasting there way into bad code, although we still need to fixup lots of the pre-existing examples.

Reply:
Did find a way to avoid it.